### PR TITLE
urls::format

### PR DESCRIPTION
**Pull request by alandefreitas**

### CODE REVIEW

File: include/boost/url/experimental/detail/impl/url_template.ipp [hunk 1]
```diff
@@ -0,0 +1,379 @@
...
+            ++it;
+            continue;
+        }
+        if (*it == '{')
+        {
+            while (
+                it != end &&
+                *it != '}')
+                ++it;
+            BOOST_ASSERT(it != end);
```
Comment (vinniefalco) [inline include/boost/url/experimental/detail/impl/url_template.ipp:125]: This assert is weird, shouldn't you write the loop this way
```
while (*it != '}' )
{
    ++it;
    BOOST_ASSERT( it != end );
}
```

File: include/boost/url/experimental/detail/vformat.hpp [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+//
+
+#ifndef BOOST_URL_EXPERIMENTAL_DETAIL_FORMAT_HPP
+#define BOOST_URL_EXPERIMENTAL_DETAIL_FORMAT_HPP
+
+#include <boost/url/detail/config.hpp>
+#include <boost/url/string_view.hpp>
+#include <boost/url/url.hpp>
+#include <boost/url/experimental/detail/format_args.hpp>
+#include <functional>
```
Comment (vinniefalco) [inline include/boost/url/experimental/detail/vformat.hpp:17]: is there way to avoid including this header?

File: include/boost/url/experimental/format.hpp [hunk 3]
```diff
@@ -0,0 +1,147 @@
...
+// Copyright (c) 2022 Alan de Freitas (alandefreitas@gmail.com)
+//
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/url
+//
+
+#ifndef BOOST_URL_EXPERIMENTAL_FORMAT_HPP
+#define BOOST_URL_EXPERIMENTAL_FORMAT_HPP
```
Comment (vinniefalco) [inline include/boost/url/experimental/format.hpp:11]: This should not be in "experimental" since it is not an experimental feature. And all new public includes need to be added to `<boost/url.hpp>`

File: include/boost/url/experimental/detail/impl/url_template.ipp [hunk 4]
```diff
@@ -0,0 +1,379 @@
...
+        n.frag);
+
+    // Apply
+    idx = 0;
+    using parts = urls::detail::parts_base;
+    if (!scheme.empty())
+    {
+        if (scheme.starts_with('{'))
+        {
+            url_base::op_t op(u);
```
Comment (vinniefalco) [inline include/boost/url/experimental/detail/impl/url_template.ipp:187]: If create and destroy an instance of `opt_t` more than once in a function, then you will get the Basic exception guarantee and not the Strong.

File: include/boost/url/experimental/detail/formatter.hpp [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        static constexpr auto m =
+            std::numeric_limits<
+                unsigned long long int>::digits10;
+        BOOST_ASSERT(n <= m + 1);
+        ignore_unused(m);
+        v = v0;
+        while (n)
+        {
+            int d = v / p;
+            encode_one(out, '0' + d, cs);
```
Comment (vinniefalco) [inline include/boost/url/experimental/detail/formatter.hpp:203]: but we don't need the character set here? for digits?

File: include/boost/url/experimental/detail/vformat.hpp (hunk 2 above)
Comment (alandefreitas) [reply include/boost/url/experimental/detail/vformat.hpp:17]: I think it's worth including it because we would be replicating most of the functionality if we didn't.

File: include/boost/url/experimental/detail/impl/url_template.ipp (hunk 4 above)
Comment (alandefreitas) [reply include/boost/url/experimental/detail/impl/url_template.ipp:187]: I'll do that anyway but a single `opt_t` won't get us a strong guarantee either, right?

File: include/boost/url/experimental/detail/formatter.hpp [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+
+template <class T>
+struct formatter;
+
+template<class CharSet>
+void
+encode_one(
+    char*& out,
+    char c,
+    CharSet const& unreserved)
```
[the remark this answers is not in the document]
Comment (alandefreitas) [reply include/boost/url/experimental/detail/formatter.hpp:41]: Is `lut_chars` more efficient than type-erasing the function?

File: include/boost/url/experimental/detail/formatter.hpp (hunk 5 above)
Comment (alandefreitas) [reply include/boost/url/experimental/detail/formatter.hpp:203]: The charset is a parameter because we don't know in what component the string will be formatted. 

Maybe we could consider we don't need to generalize the algorithm that much because digits are accepted unencoded in all components so we could bypass the `cs` here and the same could be said about other combinations of component/type/charset. 

I'm not sure removing the general algorithm is worth it though. The code might become more difficult to follow without a big benefit.

File: include/boost/url/experimental/detail/vformat.hpp (hunk 2 above)
Comment (vinniefalco) [reply include/boost/url/experimental/detail/vformat.hpp:17]: Do users end up seeing it? What do we need from it?

File: include/boost/url/experimental/detail/formatter.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (alandefreitas) [reply include/boost/url/experimental/detail/formatter.hpp:41]: `cs` is unused?

File: include/boost/url/experimental/detail/vformat.hpp (hunk 2 above)
Comment (alandefreitas) [reply include/boost/url/experimental/detail/vformat.hpp:17]: When implementing our mini-fmtlib, the format `args` are type-erased with a `handle` type that can format the string for whatever the original type was. Then I used `std::function` for that because a custom type would be 90% the same as `std::function`. 

However, because `function` is being used to type erase the formatter, they are always the same types. That means we can get rid of any extra compile time spent on instantiations with `extern template` in the `ipp` file.

File: include/boost/url/experimental/detail/formatter.hpp (hunk 6 above)
[the remark this answers is not in the document]
Comment (vinniefalco) [reply include/boost/url/experimental/detail/formatter.hpp:41]: sorry, in the digits formatter thing

File: include/boost/url/experimental/detail/vformat.hpp (hunk 2 above)
Comment (alandefreitas) [reply include/boost/url/experimental/detail/vformat.hpp:17]: I'll think about it but I'm not sure how I can hide that yet. Unless I reimplement my own version of the `function<...>` overloads I need.

Comment (klemens-morgenstern) [reply include/boost/url/experimental/detail/vformat.hpp:17]: What's wrong with `extern template` ?

File: include/boost/url/experimental/detail/impl/url_template.ipp (hunk 4 above)
Comment (vinniefalco) [reply include/boost/url/experimental/detail/impl/url_template.ipp:187]: yes it does

File: include/boost/url/experimental/detail/vformat.hpp (hunk 2 above)
Comment (vinniefalco) [reply include/boost/url/experimental/detail/vformat.hpp:17]: This can be done with a plain function pointer instead of `std::function` but also it could use an `initializer_list` to avoid the need to allocate memory. Follow me for more C++ Tips.

**Comment by alandefreitas** [comment]
note to self: 

- there are useful ideas in peter's algorithm
- we can have a single `format` function

File: include/boost/url/detail/replacement_id_rule.hpp [hunk 7]
```diff
@@ -0,0 +1,88 @@
...
+namespace boost {
+namespace urls {
+namespace detail {
+
+// replacement_field ::=  "{" [arg_id] [":" format_spec "}"
+struct replacement_id_rule_t
+{
+    using value_type = string_view;
+
+    BOOST_URL_DECL
```
Comment (vinniefalco) [inline include/boost/url/detail/replacement_id_rule.hpp:35]: is this function called from user-visible code? If not, it doesn't need the dll export.

File: include/boost/url/detail/replacement_id_rule.hpp [hunk 8]
```diff
@@ -0,0 +1,88 @@
...
+    result<value_type>
+    parse(
+        char const*& it,
+        char const* end
+    ) const noexcept;
```
Comment (vinniefalco) [inline include/boost/url/detail/replacement_id_rule.hpp:40]: no I dont like this style please fix it

File: include/boost/url/rfc/detail/impl/hier_part_rule.ipp [hunk 9]
```diff
@@ -65,7 +65,13 @@ parse(
...
         t.has_authority = true;
     }
-    if(it == end)
+    // the authority requires an absolute path
+    // or an empty path
+    if(it == end ||
+       (t.has_authority &&
+         (*it != '/' &&
+          *it != '?' &&
+          *it != '#')))
```
Comment (vinniefalco) [inline include/boost/url/rfc/detail/impl/hier_part_rule.ipp:74]: This fix needs to go into develop as its own commit?

File: include/boost/url/url_base.hpp [hunk 10]
```diff
@@ -35,6 +35,7 @@ struct any_params_iter;
 struct any_segments_iter;
 struct params_iter_impl;
 struct segments_iter_impl;
+struct url_template;
```
Comment (vinniefalco) [inline include/boost/url/url_base.hpp:38]: maybe we can rename `url_template` to `pattern`?

File: include/boost/url/detail/format_args.hpp [hunk 11]
```diff
@@ -0,0 +1,1055 @@
...
+// This file implements functions and classes to
+// type-erase format arguments.
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+// state of the format string. It basically keeps
+// track of where we are in the format string.
+class format_parse_context
```
Comment (vinniefalco) [inline include/boost/url/detail/format_args.hpp:41]: `format_pos`?

File: include/boost/url/detail/format_args.hpp [hunk 12]
```diff
@@ -0,0 +1,1055 @@
...
+        , value(v)
+    {}
+};
+
+// A type erased format argument
+class format_arg
+{
+    void const* arg_;
+    void (*measure_)( format_parse_context&, measure_context&, grammar::lut_chars const&, void const* );
+    void (*fmt_)( format_parse_context& pctx, format_context& fctx, grammar::lut_chars const&, void const* );
```
Comment (vinniefalco) [inline include/boost/url/detail/format_args.hpp:118]: Why are these lines so long

File: include/boost/url/detail/format_args.hpp [hunk 13]
```diff
@@ -0,0 +1,1055 @@
...
+    string_view name_;
+
+    template <class A>
```
Comment (vinniefalco) [inline include/boost/url/detail/format_args.hpp:121]: why is this a template?

File: include/boost/url/detail/format_args.hpp [hunk 14]
```diff
@@ -0,0 +1,1055 @@
...
+    invert(T& v, std::true_type)
+    {
+        v *= -1;
+    }
+
+    static
+    void
+    invert(T&, std::false_type)
+    {}
+};
```
Comment (vinniefalco) [inline include/boost/url/detail/format_args.hpp:1048]: 1000+ line header file, is this visible to users? Or is it only included from a .ipp file?

File: include/boost/url/detail/impl/pct_format.ipp [hunk 15]
```diff
@@ -0,0 +1,215 @@
...
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+std::size_t
+pct_vmeasure(
+    grammar::lut_chars const& cs,
+    format_parse_context& pctx,
+    measure_context& mctx)
```
Comment (vinniefalco) [inline include/boost/url/detail/impl/pct_format.ipp:23]: are these two contexts always passed together?

File: include/boost/url/detail/replacement_id_rule.hpp (hunk 7 above)
Comment (alandefreitas) [reply include/boost/url/detail/replacement_id_rule.hpp:35]: We had an error in win32 because it's called from code header files I think. I don't remember which. I'll check.

File: include/boost/url/detail/impl/url_template.ipp [hunk 16]
```diff
@@ -0,0 +1,963 @@
...
+#include <boost/url/grammar/optional_rule.hpp>
+#include <boost/url/grammar/parse.hpp>
+#include <boost/url/grammar/token_rule.hpp>
+#include <boost/url/grammar/tuple_rule.hpp>
+#include <boost/url/grammar/variant_rule.hpp>
+#include <boost/url/grammar/vchars.hpp>
+#include <boost/url/grammar/detail/charset.hpp>
+#include <boost/url/rfc/detail/fragment_part_rule.hpp>
+#include <boost/url/rfc/detail/host_rule.hpp>
+#include <boost/url/rfc/ipv4_address_rule.hpp>
```
Comment (vinniefalco) [inline include/boost/url/detail/impl/url_template.ipp:30]: this is not in alphabetical order

File: include/boost/url/detail/impl/url_template.ipp [hunk 17]
```diff
@@ -0,0 +1,963 @@
...
+{
+    return grammar::parse(
+        s, url_template_rule);
+}
+
+} // detail
+} // urls
+} // boost
+
+#endif
```
Comment (vinniefalco) [inline include/boost/url/detail/impl/url_template.ipp:963]: ugh.. another thousand-liner

File: include/boost/url/detail/replacement_id_rule.hpp [hunk 18]
```diff
@@ -0,0 +1,88 @@
...
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/url
+//
+
+#ifndef BOOST_URL_DETAIL_REPLACEMENT_ID_RULE_HPP
+#define BOOST_URL_DETAIL_REPLACEMENT_ID_RULE_HPP
+
+#include <boost/url/detail/config.hpp>
```
Comment (vinniefalco) [inline include/boost/url/detail/replacement_id_rule.hpp:13]: detail headers dont need config.hpp

File: include/boost/url/detail/replacement_id_rule.hpp [hunk 19]
```diff
@@ -0,0 +1,88 @@
...
+#include <boost/url/string_view.hpp>
+#include <boost/url/grammar/alnum_chars.hpp>
+#include <boost/url/grammar/alpha_chars.hpp>
+#include <boost/url/grammar/delim_rule.hpp>
+#include <boost/url/grammar/lut_chars.hpp>
+#include <boost/url/grammar/optional_rule.hpp>
+#include <boost/url/grammar/token_rule.hpp>
+#include <boost/url/grammar/tuple_rule.hpp>
+#include <boost/url/grammar/unsigned_rule.hpp>
+#include <boost/url/grammar/variant_rule.hpp>
```
Comment (vinniefalco) [inline include/boost/url/detail/replacement_id_rule.hpp:24]: This seems like quite a lot of includes, are you certain every one of them is necessary?

File: include/boost/url/rfc/detail/impl/hier_part_rule.ipp (hunk 9 above)
Comment (alandefreitas) [reply include/boost/url/rfc/detail/impl/hier_part_rule.ipp:74]: This minor fix was getting in the way of this PR so I fixed it. I can not squash this commit if you want. Or undo it at the end open another PR.

File: include/boost/url/detail/format_args.hpp (hunk 11 above)
Comment (pdimov) [reply include/boost/url/detail/format_args.hpp:41]: format_parse_context is the standard name.

File: include/boost/url/rfc/detail/impl/hier_part_rule.ipp (hunk 9 above)
Comment (vinniefalco) [reply include/boost/url/rfc/detail/impl/hier_part_rule.ipp:74]: Its fine to fix it in the PR and also submit a separate PR to patch it in develop, git is smart enough to elide it when you squash and rebase.

File: include/boost/url/url_base.hpp (hunk 10 above)
Comment (alandefreitas) [reply include/boost/url/url_base.hpp:38]: I'm fine with both. `pattern` seems a little generalistic but I'm assuming you don't like `url_pattern`. It's an implementation detail anyway so we can change that as much as we want.

File: include/boost/url/detail/format_args.hpp (hunk 14 above)
Comment (alandefreitas) [reply include/boost/url/detail/format_args.hpp:717]: Most content has been moved to ipp files now.

File: include/boost/url/detail/impl/pct_format.ipp (hunk 15 above)
Comment (alandefreitas) [reply include/boost/url/detail/impl/pct_format.ipp:23]: It's sometimes `format_parse_context` and `format_context` as well.

File: include/boost/url/detail/impl/url_template.ipp (hunk 17 above)
Comment (alandefreitas) [reply include/boost/url/detail/impl/url_template.ipp:963]: This one is an `.ipp` file though. No intermediary rules for the pattern are included in public headers. 

Anyway, the file contains:

- pattern_rule (172 lines)
- intermediate rules for pattern_rule (334 lines)
- function to apply a pattern to a url_base (305 lines)

The rules are quite long because they are basically implementing more complex versions of each url rule. 

To what level of granularity should I break this?

File: include/boost/url/detail/format_args.hpp [hunk 20]
```diff
@@ -0,0 +1,331 @@
...
+// Distributed under the Boost Software License, Version 1.0. (See accompanying
+// file LICENSE_1_0.txt or copy at http://www.boost.org/LICENSE_1_0.txt)
+//
+// Official repository: https://github.com/boostorg/url
+//
+
+#ifndef BOOST_URL_DETAIL_FORMAT_ARGS_HPP
+#define BOOST_URL_DETAIL_FORMAT_ARGS_HPP
+
+#include <boost/url/detail/config.hpp>
```
Comment (vinniefalco) [inline include/boost/url/detail/format_args.hpp:13]: you dont have to include config.hpp in non-public headers

**Comment by alandefreitas** [comment]
I renamed `url_template` to `pattern` as recommended in https://github.com/boostorg/url/pull/634#discussion_r1050912838 but I'm now going through https://www.rfc-editor.org/rfc/rfc6570 and it's important to note it calls it "URI templates".

**Comment by alandefreitas** [comment]
A small comparison with rfc6570

https://gist.github.com/alandefreitas/cdd2e08042d952c35cb030c4638106ff

File: doc/qbk/0.main.qbk [hunk 21]
```diff
@@ -61,17 +61,23 @@
 [def __append_to__              [link url.ref.boost__urls__string_token__return_string `string_token::return_string`]]
 
 
+[def __std_format__             [@https://en.cppreference.com/w/cpp/utility/format/format `std::format`]]
+[def __std_initializer_list__   [@https://en.cppreference.com/w/cpp/utility/initializer_list `std::initializer_list`]]
 [def __std_swap__               [@https://en.cppreference.com/w/cpp/algorithm/swap `std::swap`]]
 [def __std_string__             [@https://en.cppreference.com/w/cpp/string/basic_string `std::string`]]
 [def __std_tuple__              [@https://en.cppreference.com/w/cpp/utility/tuple `std::tuple`]]
 [def __std_get__                [@https://en.cppreference.com/w/cpp/utility/tuple/get `std::get`]]
```
Comment (vinniefalco) [inline doc/qbk/0.main.qbk:69]: not alphabetical

File: doc/qbk/0.main.qbk [hunk 22]
```diff
@@ -61,17 +61,23 @@
...
 [def __std_string_view__        [@https://en.cppreference.com/w/cpp/string/basic_string_view `std::string_view`]]
+[def __std_format_string__      [@https://en.cppreference.com/w/cpp/utility/format/basic_format_string `std::format_string`]]
```
Comment (vinniefalco) [inline doc/qbk/0.main.qbk:71]: not alphabetical

File: doc/qbk/3.1.parsing.qbk [hunk 23]
```diff
@@ -123,4 +123,65 @@ information please consult the full
 [@boost:/libs/system/doc/html/system.html#ref_resultt_e `result`]
 documentation in [@boost:/libs/system/doc/html/system.html Boost.System].
 
+[heading Formatting]
```
Comment (vinniefalco) [inline doc/qbk/3.1.parsing.qbk:126]: Why is "Formatting" in the Parsing section?

File: doc/qbk/quickref.xml [hunk 24]
```diff
@@ -80,6 +80,9 @@
         <bridgehead renderas="sect3">Functions</bridgehead>
         <simplelist type="vert" columns="1">
           <member><link linkend="url.ref.boost__urls__operator_lt__lt_">operator&lt;&lt;</link></member>
+          <member><link linkend="url.ref.boost__urls__format">format</link></member>
+          <member><link linkend="url.ref.boost__urls__format_to">format_to</link></member>
+          <member><link linkend="url.ref.boost__urls__arg">arg</link></member>
```
Comment (vinniefalco) [inline doc/qbk/quickref.xml:85]: This is not in alphabetical order and why did you add `arg`?

File: include/boost/url/detail/impl/format_args.ipp [hunk 25]
```diff
@@ -0,0 +1,642 @@
...
+    {
+        for (std::size_t i = 0; i < rpad; ++i)
+            encode_one(out, fill, cs);
+    }
+    return out;
+}
+
+} // detail
+} // urls
+} // boost
```
Comment (vinniefalco) [inline include/boost/url/detail/impl/format_args.ipp:640]: This is a medium sized file, but at least its in the .ipp and not a visible header.

File: include/boost/url/detail/impl/pattern.ipp [hunk 26]
```diff
@@ -0,0 +1,947 @@
...
+    string_view s)
+{
+    return grammar::parse(
+        s, pattern_rule);
+}
+
+} // detail
+} // urls
+} // boost
+
```
Comment (vinniefalco) [inline include/boost/url/detail/impl/pattern.ipp:946]: big file, but in the .ipp - good

File: include/boost/url/detail/impl/pattern.ipp [hunk 27]
```diff
@@ -0,0 +1,947 @@
...
+};
+
+constexpr pattern_rule_t pattern_rule{};
+
+result<pattern>
+parse_pattern(
+    string_view s)
+{
+    return grammar::parse(
+        s, pattern_rule);
```
Comment (vinniefalco) [inline include/boost/url/detail/impl/pattern.ipp:940]: This is a good function, it hides the rule declarations from the public headers

File: include/boost/url/detail/pattern.hpp [hunk 28]
```diff
@@ -0,0 +1,60 @@
...
+    bool has_pass = false;
+    bool has_port = false;
+    bool has_query = false;
+    bool has_frag = false;
+
+    BOOST_URL_DECL
+    void
+    apply(
+        url_base& u,
+        format_args const& args) const;
```
Comment (vinniefalco) [inline include/boost/url/detail/pattern.hpp:48]: Yep this all makes sense

File: include/boost/url/detail/pct_format.hpp [hunk 29]
```diff
@@ -0,0 +1,60 @@
...
+
+#include <boost/url/string_view.hpp>
+#include <boost/url/url.hpp>
+#include <boost/url/grammar/lut_chars.hpp>
+#include <boost/url/detail/format_args.hpp>
+
+// Functions to format a single fmt string with
+// pct-encoding. This is used to format
+// generalized patterns with multiple replacement
+// ids.
```
Comment (vinniefalco) [inline include/boost/url/detail/pct_format.hpp:21]: useless comment

File: include/boost/url/detail/replacement_field_rule.hpp [hunk 30]
```diff
@@ -0,0 +1,74 @@
...
+#include <boost/url/string_view.hpp>
+#include <boost/url/grammar/variant_rule.hpp>
+#include <boost/url/grammar/unsigned_rule.hpp>
+
+namespace boost {
+namespace urls {
+namespace detail {
+
+// replacement_field ::=  "{" [arg_id] [":" format_spec "}"
+struct replacement_field_rule_t
```
Comment (vinniefalco) [inline include/boost/url/detail/replacement_field_rule.hpp:23]: Do these need to be in this file or could they just go to the .ipp where they are used?

File: include/boost/url/format.hpp [hunk 31]
```diff
@@ -0,0 +1,415 @@
...
+        @ref format.
+
+*/
+inline
+void
+format_to(
+    url_base& u,
+    string_view fmt,
+#ifdef BOOST_URL_DOCS
+    std::initializer_list<__implementation_defined__> args
```
Comment (vinniefalco) [inline include/boost/url/format.hpp:365]: You could make this look like `Args const&&... args` in the docs if you want, it might be more clear than an init-list of implementation defined things. And that should probably be `__see_below__` instead of `__implementation_defined__`, because you have to document what things the user can pass for args.

File: include/boost/url/format.hpp [hunk 32]
```diff
@@ -0,0 +1,415 @@
...
+    to an implementation defined type able to
+    store the name and a reference to any type
+    potentially used as a format argument.
+
+    @par Example
+    @code
+    assert(format("user/{id}", arg("id", 1)).buffer() == "user/1");
+    @endcode
+
+    @return A type that temporarily stored a named argument
```
Comment (vinniefalco) [inline include/boost/url/format.hpp:391]: Functions can't return types....

File: include/boost/url/format.hpp [hunk 33]
```diff
@@ -0,0 +1,415 @@
...
+    std::initializer_list<detail::format_arg> args
+#endif
+    )
+{
+    detail::vformat_to(
+        u, fmt, detail::format_args(
+            args.begin(), args.end()));
+}
+
+/** Construct a named argument for a replacement field
```
Comment (vinniefalco) [inline include/boost/url/format.hpp:376]: Instead of "construct a named argument" you can say "designate a named argument"

**Comment by vinniefalco** [comment]
We can merge this as-is and continue to refine it, addressing my comments and other things. I did not check the functionality but only the physical structure of the code.

File: doc/qbk/3.1.parsing.qbk (hunk 23 above)
Comment (alandefreitas) [reply doc/qbk/3.1.parsing.qbk:126]: It can be in the Parsing section, another section after "Parsing", or in the Containers "section"? It does involve parsing the template though.

image

File: doc/qbk/quickref.xml (hunk 24 above)
Comment (alandefreitas) [reply doc/qbk/quickref.xml:85]: To explain what this public function does?

File: include/boost/url/format.hpp (hunk 31 above)
Comment (alandefreitas) [reply include/boost/url/format.hpp:365]: > You could make this look like Args const&&... args in the docs if you want

How would we differentiate that from the `Args const&&... args`?

File: doc/qbk/3.1.parsing.qbk (hunk 23 above)
Comment (vinniefalco) [reply doc/qbk/3.1.parsing.qbk:126]: it has nothing to do with parsing.

Comment (vinniefalco) [reply doc/qbk/3.1.parsing.qbk:126]: put it after Percent Encoding, as Formatting

File: doc/qbk/quickref.xml (hunk 24 above)
Comment (vinniefalco) [reply doc/qbk/quickref.xml:85]: yeah, alphabetical please though

File: include/boost/url/format.hpp (hunk 31 above)
Comment (vinniefalco) [reply include/boost/url/format.hpp:365]: oh... does the `initializer_list` overload need to be public?

Comment (alandefreitas) [reply include/boost/url/format.hpp:365]: Yes. The user uses it directly. 

It's not an implementation detail that emulates `Args const&&... args` accepting initializer_lists.

Comment (vinniefalco) [reply include/boost/url/format.hpp:365]: Why are we offering two ways to do the same thing?

File: include/boost/url/detail/replacement_field_rule.hpp (hunk 30 above)
Comment (alandefreitas) [reply include/boost/url/detail/replacement_field_rule.hpp:23]: It's kind of both. Only .ipp files include it. But it's in a separate header because many .ipp files use these rules. The router will use them too.

File: include/boost/url/format.hpp (hunk 31 above)
Comment (alandefreitas) [reply include/boost/url/format.hpp:365]: Yes. The second overload was suggested by Peter to simplify named arguments.

```cpp
url u = format("https://{username}.gigantic-server.com:{port}/{basePath}/{path}",
            arg("basePath", "v2"), arg("path", "index.html"), arg("port", 80), arg("username", "joe"));,
assert(u.buffer() == "https://joe.gigantic-server.com:80/v2/index.html");
```

becomes

```cpp
url u = format("https://{username}.gigantic-server.com:{port}/{basePath}/{path}",
            {{"basePath", "v2"}, {"path", "index.html"}, {"port", 80}, {"username", "joe"}});
assert(u.buffer() == "https://joe.gigantic-server.com:80/v2/index.html");
```

It doesn't make much sense when the arguments are not named. That's one of the discussions we had on slack.

In a way, if there's demand for that, this could converge to some kind of map concept, although that would involve changing other things. 

```cpp
url u = format("https://{username}.gigantic-server.com:{port}/{basePath}/{path}", json_obj_from_openapi);
assert(u.buffer() == "https://joe.gigantic-server.com:80/v2/index.html");
```
